Fix conversion documentation

rui-mo opened the pull request:
Fix some inaccurate descriptions found in conversion.rst. Document the 
limitation of cast from string to real. Add unit tests for cast from float to 
bool, cast from string to float, and cast from timestamp with timezone to
timestamp.

rui-mo commented:
@mbasmanova Could you take a look? Thanks.

rui-mo commented on velox/docs/functions/presto/conversion.rst:
```diff
@@ -428,6 +429,8 @@ Valid examples
   SELECT cast('1.' as real); -- 1.0
   SELECT cast('1' as real); -- 1.0
+  SELECT cast('1.2f' as real); -- 1.2
+  SELECT cast('1.2d' as real); -- 1.2
```
Same.

rui-mo replied:
They are supported in Velox. Document the limitation.

rui-mo commented on velox/docs/functions/presto/conversion.rst:
```diff
@@ -363,6 +364,8 @@ Valid examples
   SELECT cast(infinity() as boolean); -- true
   SELECT cast(0.0000000000001 as boolean); -- true
+  SELECT cast(0.5 as boolean); -- true
+  SELECT cast(-0.5 as boolean); -- true
```
This is supported by Presto and Spark. Will check if Velox has aligned result.

rui-mo commented on velox/docs/functions/presto/conversion.rst:
```diff
@@ -448,8 +450,22 @@ consistent with other supported cases of cast.
 
 ::
 
+  SELECT cast('infinity' as real); -- Infinity
```
presto:test> SELECT cast('infinity' as real);
Query 20231120_033856_00032_a9j2g failed: Cannot cast 'infinity' to REAL

rui-mo replied:
Same as https://github.com/facebookincubator/velox/pull/7629/files#r1408606756. Do I have anything missing?

rui-mo commented on velox/docs/functions/presto/conversion.rst:
```diff
@@ -448,8 +450,22 @@ consistent with other supported cases of cast.
 
+  SELECT cast('infinity' as real); -- Infinity
+  SELECT cast('-infinity' as real); -- -Infinity
+  SELECT cast('inf' as real); -- Infinity
```
presto:test> SELECT cast('inf' as real);
Query 20231120_033915_00034_a9j2g failed: Cannot cast 'inf' to REAL

rui-mo commented on velox/docs/functions/presto/conversion.rst:
```diff
@@ -448,8 +450,22 @@ consistent with other supported cases of cast.
+  SELECT cast('-infinity' as real); -- -Infinity
+  SELECT cast('inf' as real); -- Infinity
   SELECT cast('InfiNiTy' as real); -- Infinity
+  SELECT cast('INFINITY' as real); -- Infinity
```
presto:test> SELECT cast('INFINITY' as real);
Query 20231120_025230_00010_a9j2g failed: Cannot cast 'INFINITY' to REAL

rui-mo commented on velox/docs/functions/presto/conversion.rst:
```diff
@@ -448,8 +450,22 @@ consistent with other supported cases of cast.
   SELECT cast('InfiNiTy' as real); -- Infinity
+  SELECT cast('INFINITY' as real); -- Infinity
   SELECT cast('nAn' as real); -- NaN
+  SELECT cast('nan' as real); -- NaN
```
presto:test> SELECT cast('nan' as real);
Query 20231120_033942_00035_a9j2g failed: Cannot cast 'nan' to REAL

rui-mo reviewed (commented):
These throw in Presto, but succeed in Velox. Document them in the differences part.

rui-mo commented:
@gggrace14 @kagamiori This PR fixes some inaccurate descriptions in cast documentation. Could you spare some time to review? Thanks.

kagamiori commented on velox/docs/functions/presto/conversion.rst:
```diff
@@ -287,7 +287,7 @@ Valid examples if cast_to_int_by_truncate=true
 ::
 
-  SELECT cast('12345.67' as tinyint); -- 12345
+  SELECT cast('12345.67' as bigint); -- 12345
```
Nice catch!

kagamiori reviewed (commented):
Hi @rui-mo, thank you for fixing the documentation of conversion! I left a few comments. Also, since you updated the documentation about casting TIMESTAMP WITH TIME ZONE to TIMESTAMP, would you also like to add test cases for that?

rui-mo commented on velox/docs/functions/presto/conversion.rst:
```diff
@@ -429,15 +432,14 @@ Valid examples
-  SELECT cast('infinity' as real); -- Infinity (case insensitive)
-  SELECT cast('-infinity' as real); -- -Infinity (case insensitive)
-  SELECT cast('nan' as real); -- NaN (case insensitive)
+  SELECT cast('Infinity' as real); -- Infinity (case insensitive)
+  SELECT cast('-Infinity' as real); -- -Infinity (case insensitive)
+  SELECT cast('NaN' as real); -- NaN (case insensitive)
```
Thanks for your review. I got failure on Presto for the original examples, but they are supported in Velox indeed. I moved them from the right examples to the inconsistent behaviors of Presto and Velox.

![image](https://github.com/facebookincubator/velox/assets/41687378/1ed2b109-73ff-42c0-aa5c-6bc84a963519)

kagamiori replied:
OK, this makes sense to me. Thank you for the explanation!

rui-mo commented:
> Also, since you updated the documentation about casting TIMESTAMP WITH TIME ZONE to TIMESTAMP, would you also like to add test cases for that?

@kagamiori Thanks for your review. Looks like these tests already exist for enabling and disabling this config, but the documentation reverses their semantics.

https://github.com/facebookincubator/velox/blob/fc9fb37e3090d1c553c8ddbd056cd2abc45e744f/velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp#L3650

https://github.com/facebookincubator/velox/blob/fc9fb37e3090d1c553c8ddbd056cd2abc45e744f/velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp#L3670

kagamiori commented:
> > Also, since you updated the documentation about casting TIMESTAMP WITH TIME ZONE to TIMESTAMP, would you also like to add test cases for that?
> 
> @kagamiori Thanks for your review. Looks like these tests already exist for enabling and disabling this config, but the documentation reverses their semantics.
> 
> https://github.com/facebookincubator/velox/blob/fc9fb37e3090d1c553c8ddbd056cd2abc45e744f/velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp#L3650
> 
> https://github.com/facebookincubator/velox/blob/fc9fb37e3090d1c553c8ddbd056cd2abc45e744f/velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp#L3670

I think this test is for casting DATE to TIMESTAMP. The unit test for casting TIMESTAMP WITH TIME ZONE to TIMESTAMP doesn't test both configs I think.
https://github.com/facebookincubator/velox/blob/ffd2ee0c711b7015eff015f98ee331b8ed8ae917/velox/functions/prestosql/tests/TimestampWithTimeZoneCastTest.cpp#L49

@zacw7, would it make sense to add a test case to TimestampWithTimeZoneCastTest.cpp for adjust_timestamp_to_session_timezone being true?

zacw7 commented on velox/docs/functions/presto/conversion.rst:
```diff
@@ -560,28 +576,28 @@ From TIMESTAMP WITH TIME ZONE
-"1970-01-01 00:00:00.000 America/Los_Angeles" becomes "1970-01-01 00:00:00.000".
-
-Otherwise, timestamp is shifted by the offset of the timezone. For example,
+If set to true, timestamp is shifted by the offset of the timezone. For example,
```
Wondering why do we want to make this change? Based on the current logic, only when `adjust_timestamp_to_session_timezone == false`, the adjustment would be made right? See [code ref](https://fburl.com/hy0qw0m7).

rui-mo commented:
> Based on the current logic, only when adjust_timestamp_to_session_timezone == false, the adjustment would be made right?

@zacw7 @kagamiori Thank you for the review. Added test cases to for `adjust_timestamp_to_session_timezone` being true. It turns out the documentation matches with the actual behavior, so I removed this change.
But I also notice for cast date to timestamp, when `adjust_timestamp_to_session_timezone == false`, the timestamp is returned as is (please correct me if I am wrong). This may causes the behaviors controlled by the same config to be opposite.
https://github.com/facebookincubator/velox/blob/593309da9de650c0ac5d098bc9e99c3513b61b40/velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp#L3670-L3671
Is there any special concern? Glad to hear your opinions, thanks!

kagamiori commented:
> > Based on the current logic, only when adjust_timestamp_to_session_timezone == false, the adjustment would be made right?
> 
> @zacw7 @kagamiori Thank you for the review. Added test cases to for `adjust_timestamp_to_session_timezone` being true. It turns out the documentation matches with the actual behavior, so I removed this change. But I also notice for cast date to timestamp, when `adjust_timestamp_to_session_timezone == false`, the timestamp is returned as is (please correct me if I am wrong). This may causes the behaviors controlled by the same config to be opposite.
> 
> https://github.com/facebookincubator/velox/blob/593309da9de650c0ac5d098bc9e99c3513b61b40/velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp#L3670-L3671
> 
> 
> Is there any special concern? Glad to hear your opinions, thanks!

Hi @rui-mo, I looked at the code of casting TimestampWithTimezone from/to Timestamp and I think you were right that the documentation is reversed compared to the actual behavior. So the behavior of cast Date to Timestamp is correct. Let me double check with @zacw7 who is the original author of this function. I'll get back to you tomorrow.

rui-mo commented:
@kagamiori @zacw7 Thanks a lot.

zacw7 commented:
Thanks for working on this. I'm taking a closer look on this.

zacw7 commented:
I can confirm that current cast behaviors are exactly the same as in Presto.

For cast _timestamp with time zone_ to _timestamp_. Same results can be observed in both Presto and Velox:
```SQL
SET SESSION legacy_timestamp=true; -- adjustTimestampToTimezone == true
SELECT cast(from_unixtime(0, '+06:00') as timestamp); -- 1969-12-31 16:00:00.000 (unixtime: 0)

SET SESSION legacy_timestamp=false; -- adjustTimestampToTimezone == false
SELECT cast(from_unixtime(0, '+06:00') as timestamp); -- 1970-01-01 06:00:00.000 (unixtime: 21600000)
```
So yes, the current documentation matches this behavior: If _adjust_timestamp_to_session_timezone_  is set to true, input timezone is ignored and timestamp is returned as is.

For cast date to timestamp, unfortunately this is how Presto currently behaves. https://github.com/prestodb/presto/blob/04138bc26a4fc801630fe936463d89d071d00c0a/presto-main/src/main/java/com/facebook/presto/type/DateOperators.java#L117-L127

kagamiori commented:
> I can confirm that current cast behaviors are exactly the same as in Presto.
> 
> For cast _timestamp with time zone_ to _timestamp_. Same results can be observed in both Presto and Velox:
> 
> ```sql
> SET SESSION legacy_timestamp=true; -- adjustTimestampToTimezone == true
> SELECT cast(from_unixtime(0, '+06:00') as timestamp); -- 1969-12-31 16:00:00.000 (unixtime: 0)
> 
> SET SESSION legacy_timestamp=false; -- adjustTimestampToTimezone == false
> SELECT cast(from_unixtime(0, '+06:00') as timestamp); -- 1970-01-01 06:00:00.000 (unixtime: 21600000)
> ```
> 
> So yes, the current documentation matches this behavior: If _adjust_timestamp_to_session_timezone_ is set to true, input timezone is ignored and timestamp is returned as is.
> 
> For cast date to timestamp, unfortunately this is how Presto currently behaves. https://github.com/prestodb/presto/blob/04138bc26a4fc801630fe936463d89d071d00c0a/presto-main/src/main/java/com/facebook/presto/type/DateOperators.java#L117-L127

Hi @zacw7, thank you very much for confirming the Presto behavior! I have a question about the documentation.

The current Velox documentation says
```
-- `adjust_timestamp_to_session_timezone` is true
SELECT cast(timestamp '1970-01-01 00:00:00 America/Los_Angeles' as timestamp); -- 1970-01-01 00:00:00.000
```

Isn't this the same behavior as your example of `SET SESSION legacy_timestamp=false` above?
> SET SESSION legacy_timestamp=false; -- adjustTimestampToTimezone == false
> SELECT cast(from_unixtime(0, '+06:00') as timestamp); -- 1970-01-01 06:00:00.000 (unixtime: 21600000)

On the other hand, the Velox documentation also says
```
-- `adjust_timestamp_to_session_timezone` is true
...
SELECT cast(from_unixtime(0, '+06:00') as timestamp); -- 1970-01-01 00:00:00.000
```

This does match your example above. I feel like there is a mismatch of examples in the current Velox documentation. What do you think?

rui-mo commented:
@zacw7 @kagamiori Thank you for confirming. Added the test examples on `adjust_timestamp_to_timezone`, could you help review again?

zacw7 commented:
> Because from_unixtime(0, '+06:00') is 1970-01-01 06:00:00.000 +06:00.
I think `from_unixtime(0, '+06:00')` should be  `1970-01-01 00:00:00.000 +06:00`

zacw7 commented:
Given the fact that `legacy_timestamp/adjust_timestamp_to_session_timezone` also controls how timestamp displays as output and the behavior difference may be hard to observe in the format of timestamp. I would suggest revise the examples and print the output as unixtime:

```SQL
  -- `adjust_timestamp_to_session_timezone` is true
  SELECT to_unixtime(cast(timestamp '1970-01-01 00:00:00 America/Los_Angeles' as timestamp)); -- 28,800 (1970-01-01 00:00:00)
  SELECT to_unixtime(cast(timestamp '2012-03-09 10:00:00 Asia/Chongqing' as timestamp)); -- 1,331,258,400 (2012-03-09 02:00:00)
  SELECT to_unixtime(cast(from_unixtime(0, '+06:00') as timestamp)); -- 0 (1970-01-01 00:00:00)
  SELECT to_unixtime(cast(from_unixtime(0, '-02:00') as timestamp)); -- 0 (1970-01-01 00:00:00)

  -- `adjust_timestamp_to_session_timezone` is false
  SELECT to_unixtime(cast(timestamp '1970-01-01 00:00:00 America/Los_Angeles' as timestamp)); -- 0 (1970-01-01 00:00:00)
  SELECT to_unixtime(cast(timestamp '2012-03-09 10:00:00 Asia/Chongqing' as timestamp)); -- 1,331,287,200 (2012-03-09 10:00:00)
  SELECT to_unixtime(cast(from_unixtime(0, '+06:00') as timestamp)); -- 21,600 (1970-01-01 06:00:00)
  SELECT to_unixtime(cast(from_unixtime(0, '-02:00') as timestamp)); -- -7,200 (1969-12-31 22:00:00)
```

zacw7 reviewed (changes requested):
Given the fact that `legacy_timestamp/adjust_timestamp_to_session_timezone` also controls how timestamp displays as output and the behavior difference may be hard to observe in the format of timestamp. I would suggest revise the examples and print the output as unixtime:

(Examples in section `From TIMESTAMP WITH TIME ZONE`).
```SQL
  -- `adjust_timestamp_to_session_timezone` is true
  SELECT to_unixtime(cast(timestamp '1970-01-01 00:00:00 America/Los_Angeles' as timestamp)); -- 28,800 (1970-01-01 08:00:00)
  SELECT to_unixtime(cast(timestamp '2012-03-09 10:00:00 Asia/Chongqing' as timestamp)); -- 1,331,258,400 (2012-03-09 02:00:00)
  SELECT to_unixtime(cast(from_unixtime(0, '+06:00') as timestamp)); -- 0 (1970-01-01 00:00:00)
  SELECT to_unixtime(cast(from_unixtime(0, '-02:00') as timestamp)); -- 0 (1970-01-01 00:00:00)

  -- `adjust_timestamp_to_session_timezone` is false
  SELECT to_unixtime(cast(timestamp '1970-01-01 00:00:00 America/Los_Angeles' as timestamp)); -- 0 (1970-01-01 00:00:00)
  SELECT to_unixtime(cast(timestamp '2012-03-09 10:00:00 Asia/Chongqing' as timestamp)); -- 1,331,287,200 (2012-03-09 10:00:00)
  SELECT to_unixtime(cast(from_unixtime(0, '+06:00') as timestamp)); -- 21,600 (1970-01-01 06:00:00)
  SELECT to_unixtime(cast(from_unixtime(0, '-02:00') as timestamp)); -- -7,200 (1969-12-31 22:00:00)
```

kagamiori commented:
Thanks @zacw7 for diligently looking into the behaviors of Presto legacy_timestamp and Velox adjust_timestamp_to_session_timezone and suggesting the proper clarification to the Velox documentation.

@rui-mo, in addition to the changes suggested by @zacw7 above, could you also change the paragraphs above the those examples to the following:
```
From TIMESTAMP WITH TIME ZONE
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The results depend on whether configuration property `adjust_timestamp_to_session_timezone` is set or not.

If set to true, input timezone is ignored and timestamp is returned as is. For example,
"1970-01-01 00:00:00.000 America/Los_Angeles" becomes "1970-01-01 08:00:00".

Otherwise, timestamp is shifted by the offset of the timezone. For example,
"1970-01-01 00:00:00.000 America/Los_Angeles" becomes "1970-01-01 00:00:00".
```

This change makes the examples here match the changes @zacw7 suggested.

rui-mo commented:
@zacw7 @kagamiori Many thanks for taking efforts to provide these great examples and suggestions. Just changed the example results to the data format got from Presto.
```
presto:test> SELECT to_unixtime(cast(timestamp '2012-03-09 10:00:00 Asia/Chongqing' as timestamp));
    _col0    
-------------
 1.3312872E9 
(1 row)
```

zacw7 reviewed (approved):
LGTM

kagamiori reviewed (approved):
LGTM. Thanks a lot for iterating on it!

kagamiori commented:
@rui-mo, looks like the build on the base commit was broken. Could you rebase to the latest main?

rui-mo commented:
@kagamiori Rebased. Thanks for noticing.
